File: PathoFroh/src/main/java/com/patho/main/action/handler/DialogReturnHandler.java
package com.patho.main.action.handler;

import com.patho.main.action.dialog.diagnosis.DiagnosisPhaseExitDialog.DiagnosisPhaseExitData;
import com.patho.main.action.dialog.miscellaneous.DeleteDialog.DeleteEvent;
import com.patho.main.action.dialog.patient.MergePatientDialog.PatientMergeEvent;
import com.patho.main.action.dialog.settings.UserSettingsDialog.ReloadUserEvent;
import com.patho.main.action.dialog.slides.StainingPhaseExitDialog.StainingPhaseExitData;
import com.patho.main.action.dialog.worklist.WorklistSearchDialog.WorklistSearchReturnEvent;
import com.patho.main.model.patient.*;
import com.patho.main.service.BlockService;
import com.patho.main.service.SampleService;
import com.patho.main.service.SlideService;
import com.patho.main.service.UserService;
import com.patho.main.util.dialogReturn.PatientReturnEvent;
import com.patho.main.util.dialogReturn.ReloadEvent;
import com.patho.main.util.dialogReturn.ReloadTaskEvent;
import com.patho.main.util.event.dialog.StainingPhaseUpdateEvent;
import org.primefaces.event.SelectEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Controller;

@Controller
@Scope("session")
public class DialogReturnHandler {

    protected final Logger logger = LoggerFactory.getLogger(this.getClass());

    @Autowired
    private WorklistHandler worklistHandler;

    @Autowired
    private WorkPhaseHandler workPhaseHandler;

    @Autowired
    private UserService userService;

    @Autowired
    private SlideService slideService;

    @Autowired
    private BlockService blockService;

    @Autowired
    private SampleService sampleService;

    @Autowired
    private CentralHandler centralHandler;

    /**
     * Dialog return function, reloads the selected task
     *
     * @param event
     */
    public void onDefaultDialogReturn(SelectEvent event) {
        if (event.getObject() != null) {
            // Patient reload event
            if (event.getObject() instanceof PatientReturnEvent) {
                logger.debug("Patient add event reload event.");
                if (((PatientReturnEvent) event.getObject()).getTask() != null)
                    worklistHandler.getCurrent().setSelectedTask(((PatientReturnEvent) event.getObject()).getTask());
                worklistHandler.getCurrent().reloadSelectedPatientAndTask();

                centralHandler.loadViews(CentralHandler.Load.RELOAD_TASK_STATUS);
                // staining phase reload event
            } else if (event.getObject() instanceof ReloadTaskEvent || event.getObject() instanceof ReloadEvent) {
                logger.debug("Task reload event.");
                if (event.getObject() instanceof ReloadTaskEvent
                        && ((ReloadTaskEvent) event.getObject()).getTask() != null)
                    worklistHandler.replaceTaskInWorklist(((ReloadTaskEvent) event.getObject()).getTask());
                else
                    worklistHandler.replaceTaskInWorklist();
            }
        }
    }


    public void onStatingPhaseExitReturn(SelectEvent event) {
        if (event.getObject() instanceof StainingPhaseExitData) {
            logger.debug("Staining phase exit dialog return");
            StainingPhaseExitData data = (StainingPhaseExitData) event.getObject();

            workPhaseHandler.endStainingPhase(data.getTask(), data.isEndStainingPhase(),
                    data.isRemoveFromStainingList(), data.isGoToDiagnosisPhase(), data.isRemoveFromWorklist());

            worklistHandler.replaceTaskInWorklist();

            return;
        }
        onDefaultDialogReturn(event);
    }

    public void onDiagnosisPhaseExitReutrn(SelectEvent event) {
        if (event.getObject() instanceof DiagnosisPhaseExitData) {
            logger.debug("Diagnosis phase exit dialog return");

            DiagnosisPhaseExitData data = (DiagnosisPhaseExitData) event.getObject();
            workPhaseHandler.endDiagnosisPhase(data.getTask(),
                    data.isAllRevisions() ? null : data.getSelectedRevision(), data.isRemoveFromDiangosisList(),
                    data.isNotification(), data.isRemoveFromWorklist());

            worklistHandler.replaceTaskInWorklist();
            return;
        }
        onDefaultDialogReturn(event);
    }

}

File: PathoFroh/src/main/kotlin/com/patho/main/action/handler/DialogReturnHandler.kt
package com.patho.main.action.handler

import com.patho.main.model.patient.Block
import com.patho.main.model.patient.Sample
import com.patho.main.model.patient.Slide
import com.patho.main.model.patient.Task
import com.patho.main.service.BlockService
import com.patho.main.service.SampleService
import com.patho.main.service.SlideService
import com.patho.main.service.UserService
import com.patho.main.util.dialogReturn.ReloadTaskEvent
import com.patho.main.util.event.dialog.*
import org.primefaces.event.SelectEvent
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

@Component
@Scope(value = "session")
open class DialogReturnHandler @Autowired constructor(
        private val worklistHandler: WorklistHandler,
        private val workPhaseHandler: WorkPhaseHandler,
        private val slideService: SlideService,
        private val blockService: BlockService,
        private val sampleService: SampleService,
        private val userService: UserService,
        private val centralHandler: CentralHandler) : AbstractHandler() {
    override fun loadHandler() {
    }


    open fun onDefaultReturn(event: SelectEvent) {
        logger.debug("Default Return Handler")
        event ?: return;

        val obj = event.`object`

        when (obj) {
            // Patient reload event
            //TODO Test CreateTaskDialog
            //TODO Test SearchPatientDialog
            is PatientReloadEvent -> {
                logger.debug("Patient reload event")

                if (obj.select) {
                    if (obj.task != null)
                        centralHandler.onSelectTaskAndPatient(obj.task)
                    else if (obj.patient != null)
                        centralHandler.onSelectPatient(obj.patient)
                    else
                        centralHandler.onSelectPatient(worklistHandler.current.selectedPatient)
                } else {
                    if (obj.task != null)
                        worklistHandler.replaceTaskInWorklist(obj.task)
                    else if (obj.patient != null) {
                        if (worklistHandler.isSelected(obj.patient.tasks))
                            worklistHandler.replaceTaskInWorklist()
                        else
                            worklistHandler.replacePatientInWorklist(obj.patient)
                    } else {
                        if (worklistHandler.selectedTask != null)
                            worklistHandler.replaceTaskInWorklist()
                        else
                            worklistHandler.replacePatientInWorklist()
                    }
                }
            }
            // Task reload event, optional replacing task
            //Todo Test CreateDiagnosisRevisionDialog
            is ReloadTaskEvent -> {
                logger.debug("Task reload event")
                if (obj is Task)
                    worklistHandler.replaceTaskInWorklist(obj)
                else
                    worklistHandler.replaceTaskInWorklist()
            }
        }
    }

    open fun onCreateTaskReturn(event: SelectEvent) {

    }

    /**
     * Return handler for the create Sample- and Slide-Dialog
     */
    open fun onStainingPhaseUpdateReturn(event: SelectEvent) {
        if (event.`object` is StainingPhaseUpdateEvent) {
            logger.debug("Staining phase update event")
            val t = (event.`object` as StainingPhaseUpdateEvent).task
            workPhaseHandler.updateStainingPhase(t)
            worklistHandler.replaceTaskInWorklist(t)
            return
        }
        onDefaultReturn(event)
    }

    /**
     * Return handler for patient merge dialog
     */
    open fun onPatientMergeReturn(event: SelectEvent) {
        val obj = event.`object`
        if (obj is PatientMergeEvent) {
            logger.debug("Patient merge event")
            if (obj.source.archived)
                worklistHandler.removePatientFromWorklist(obj.source)
            else
                worklistHandler.replacePatientInWorklist(obj.source)

            if (obj.target.archived)
                worklistHandler.removePatientFromWorklist(obj.target)
            else
                worklistHandler.replacePatientInWorklist(obj.target)
        }
        onDefaultReturn(event)
    }

    /**
     * Return handler for DeleteDialog
     */
    open fun onDeleteTaskEntityReturn(event: SelectEvent) {
        val obj = event.`object`
        if (obj is TaskEntityDeleteEvent) {
            logger.debug("Deleting task entity event")

            val task: Task? = when (obj.obj) {
                is Slide -> slideService.deleteSlide(obj.obj,true)
                is Block -> blockService.deleteBlockAndPersist(obj.obj)
                is Sample -> sampleService.deleteSampleAndPersist(obj.obj)
                else -> null
            }

            // updating staining phase if task is not null
            if (task != null)
                onStainingPhaseUpdateReturn(SelectEvent(event.component, event.behavior, StainingPhaseUpdateEvent(task)))

            return
        }
        onDefaultReturn(event)
    }

    /**
     * Return handler for UserSettingsDialog
     */
    open fun onUserSettingsReturn(event: SelectEvent) {
        if (event.`object` is UserReloadEvent) {
            logger.debug("User settings return event")
            userService.reloadCurrentUser()
            return
        }
        onDefaultReturn(event)
    }


    open fun onSettingsReturn(event: SelectEvent) {
        if (event.`object` is SettingsReloadEvent) {
            logger.debug("Settings return event")
            centralHandler.reloadAllData()
            return
        }
        onDefaultReturn(event)
    }

    /**
     * Return event handler for selecting new worklists
     */
    open fun onWorklistSelectReturn(event: SelectEvent) {
        val obj = event.`object`
        if (obj is WorklistSelectEvent) {
            worklistHandler.addWorklist(obj.obj, true)
            return
        }
        onDefaultReturn(event)
    }

    /**
     * Event handler for adding a patient to the current worklist
     */
    open fun onSearchForPatientReturn(event: SelectEvent) {
        val obj = event.`object`
        if (obj is PatientSelectEvent) {
            logger.debug("Patient select event")

            if (obj.obj == null) {
                logger.debug("No Patient selected")
                return
            }

            worklistHandler.addPatientToWorkList(obj.obj, true, true)
            return
        }
        onDefaultReturn(event)
    }

    /**
     * Event handler for Patient delete dialog
     */
    open fun onPatientDeleteReturn(event: SelectEvent) {
        val obj = event.`object`
        if (obj is PatientDeleteEvent) {
            // TODO implement logic
            return
        }
        onDefaultReturn(event)
    }

    open fun onTaskPhaseChangeEvent(event: SelectEvent){
        val obj = event.`object`
        if (obj is RemovePatientFromWorklistEvent) {
           worklistHandler.removePatientFromWorklist(obj.obj)
            return
        }

        onDefaultReturn(event)
    }
}